numba/targets/hashing.py

```diff
@@ -0,0 +1,605 @@
+# https://github.com/python/cpython/blob/d1dd6be613381b996b9071443ef081de8e5f3aff/Objects/tupleobject.c#L347-L369
+@register_jitable(locals={'x': _Py_uhash_t,
+                          'y': _Py_hash_t,
+                          'mult': _Py_uhash_t,
+                          'l': _Py_hash_t, })
+def _tuple_hash(tup):
+    tl = len(tup)
+    mult = _PyHASH_MULTIPLIER
+    x = _Py_uhash_t(0x345678)
+    # in C this is while(--l >= 0), i is indexing tup instead of *tup++
+    for i, l in enumerate(range(tl - 1, -1, -1)):
+        y = hash(tup[i])
+        xxory = (x ^ y)
+        x = xxory * mult
+        mult += _Py_hash_t((_Py_uhash_t(82520) + l + l))
+    x += _Py_uhash_t(97531)
+    return process_return(x)
+
+# This is an obfuscated translation of CPython's tuplehash:
+# https://github.com/python/cpython/blob/d1dd6be613381b996b9071443ef081de8e5f3aff/Objects/tupleobject.c#L347-L369
+# The obfuscation occurs for a heterogeneous tuple as each tuple member needs
+# a potentially different hash() function calling for it. This cannot be done at
+# runtime as there's no way to iterate a heterogeneous tuple, so this is
+# achieved by essentially unrolling the loop over the members and inserting a
+# per-type hash function call for each member, and then simply computing the
+# hash value in an inlined/rolling fashion.
+
+
+@intrinsic
+def _tuple_hash_resolve(tyctx, val):
+    def impl(cgctx, builder, signature, args):
+        typingctx = cgctx.typing_context
+        fnty = typingctx.resolve_value_type(hash)
+        tupty, = signature.args
+        tup, = args
+        lty = cgctx.get_value_type(signature.return_type)
+        x = ir.Constant(lty, 0x345678)
+        mult = ir.Constant(lty, _PyHASH_MULTIPLIER)
+        shift = ir.Constant(lty, 82520)
+        tl = len(tupty)
+        for i, packed in enumerate(zip(tupty.types, range(tl - 1, -1, -1))):
+            ty, l = packed
+            sig = fnty.get_call_type(tyctx, (ty,), {})
+            impl = cgctx.get_function(fnty, sig)
+            tuple_val = builder.extract_value(tup, i)
+            y = impl(builder, (tuple_val,))
+            xxory = builder.xor(x, y)
+            x = builder.mul(xxory, mult)
+            lconst = ir.Constant(lty, l)
+            mult = builder.add(mult, shift)
+            mult = builder.add(mult, lconst)
+            mult = builder.add(mult, lconst)
+        x = builder.add(x, ir.Constant(lty, 97531))
+        return x
+    sig = _Py_hash_t(val)
+    return sig, impl
+
+
+@overload_method(types.BaseTuple, '__hash__')
+def tuple_hash(val):
+    if isinstance(val, types.Sequence):
+        def impl(val):
+            return _tuple_hash(val)
+        return impl
+    else:
+        def impl(val):
+            hashed = _Py_hash_t(_tuple_hash_resolve(val))
+            return process_return(hashed)
+        return impl
+
+
+# ------------------------------------------------------------------------------
+# String/bytes hashing needs hashseed info, this is from:
```

Comment:
I believe the section of the code has to do with `PYTHONHASHSEED`.  Can you add some developer notes here about that?

Reply:
Yes, done. Also, developer docs section has some information on this.